src/login/pam_kernel_keyring.c

```diff
@@ -0,0 +1,102 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
+#include <security/_pam_macros.h>
+#include <security/pam_ext.h>
+#include <security/pam_misc.h>
+#include <security/pam_modules.h>
+#include <security/pam_modutil.h>
+
+#include "keyring-util.h"
+#include "macro.h"
+#include "missing_syscall.h"
+#include "nulstr-util.h"
+#include "pam-util.h"
+#include "strv.h"
+
+/* By default, this module retieves the key stored by systemd-cryptsetup.
+ * This can be overridden by the keyname= parameter. */
+static const char* DEFAULT_KEYNAME = "cryptsetup";
+
+_public_ int pam_sm_authenticate(
+                pam_handle_t  *handle,
+                int            flags,
+                int            argc, const char   **argv) {
+
+        assert(handle);
+
+        /* Parse argv. */
+
+        assert(argc >= 0);
+        assert(argc == 0 || argv);
+
+        const char *keyname = DEFAULT_KEYNAME;
+        bool debug = false;
+
+        for (int i = 0; i < argc; i++) {
+                const char *p;
+
+                if ((p = startswith(argv[i], "keyname=")))
+                        keyname = p;
+                else if (streq(argv[i], "debug"))
+                        debug = true;
+                else
+                        pam_syslog(handle, LOG_WARNING, "Unknown parameter '%s', ignoring.", argv[i]);
+        }
+
+        pam_debug_syslog(handle, debug, "pam-kernel-keyring initializing");
+
+        /* Retrive the key. */
+
+        key_serial_t serial;
+        serial = request_key("user", keyname, NULL, 0);
+        if (serial < 0) {
+                if (errno == ENOKEY) {
+                        pam_debug_syslog(handle, debug, "Key not found: %s", keyname);
+                        return PAM_AUTHINFO_UNAVAIL;
+                } else if (errno == EKEYEXPIRED) {
+                        pam_debug_syslog(handle, debug, "Key expired: %s", keyname);
+                        return PAM_AUTHINFO_UNAVAIL;
+                } else {
+                        return pam_syslog_errno(handle, LOG_ERR, errno, "Failed to look up the key: %m");
+                }
+        }
+
+        _cleanup_(erase_and_freep) void *p = NULL;
+        _cleanup_(strv_free_erasep) char **passwords = NULL;
```

Comment:
you could move this line down, just before the strv_parse_nulst(), but doesn't matter